src/server/hotkey_collector.cpp

```diff
@@ -17,21 +17,53 @@
 
 #include "hotkey_collector.h"
 
+#include <dsn/dist/replication/replication_enums.h>
 #include <dsn/utility/smart_pointers.h>
 #include "base/pegasus_key_schema.h"
+#include <dsn/dist/fmt_logging.h>
 
 namespace pegasus {
 namespace server {
 
-hotkey_collector::hotkey_collector()
-    : _internal_collector(dsn::make_unique<hotkey_empty_data_collector>())
+hotkey_collector::hotkey_collector(dsn::replication::hotkey_type::type hotkey_type,
+                                   dsn::replication::replica_base *r_base)
+    : replica_base(r_base),
+      _state(hotkey_collector_state::STOPPED),
+      _hotkey_type(hotkey_type),
+      _internal_collector(std::make_shared<hotkey_empty_data_collector>())
 {
 }
 
-// TODO: (Tangyanzhao) implement these functions
 void hotkey_collector::handle_rpc(const dsn::replication::detect_hotkey_request &req,
                                   dsn::replication::detect_hotkey_response &resp)
 {
+    switch (req.action) {
+    case dsn::replication::detect_action::START:
+        if (is_detecting_now()) {
```

Comment:
you can encapsulate into a new function for each case.

Reply:
According to this comment https://github.com/apache/incubator-pegasus/pull/621#discussion_r506004399, I combined these functions